src/states/common/base.rs

```diff
@@ -224,53 +221,45 @@ pub trait Base: Display {
     fn try_handle_message(
         &mut self,
         sender: Option<SocketAddr>,
-        msg: MessageWithBytes,
+        mut msg_with_bytes: MessageWithBytes,
         outbox: &mut dyn EventBox,
     ) -> Result<Transition> {
-        trace!(
-            "{} - try handle message {:?}",
-            self,
-            HexFmt(msg.full_crypto_hash())
-        );
+        trace!("{} - try handle message {:?}", self, msg_with_bytes);
 
-        if !self.filter_incoming_message(&msg) {
-            trace!(
-                "{} - Known message {:?}, not handling further",
-                self,
-                HexFmt(msg.full_crypto_hash())
-            );
+        self.try_relay_message(sender, &msg_with_bytes)?;
 
+        if !self.in_dst_location(msg_with_bytes.message_dst()) {
             return Ok(Transition::Stay);
         }
 
-        self.handle_filtered_message(sender, msg, outbox)
-    }
-
-    fn handle_filtered_message(
-        &mut self,
-        sender: Option<SocketAddr>,
-        mut msg: MessageWithBytes,
-        outbox: &mut dyn EventBox,
-    ) -> Result<Transition> {
-        self.try_relay_message(&msg)?;
-
-        if !self.in_dst_location(msg.message_dst()) {
+        if self.is_message_handled(&msg_with_bytes) {
+            trace!(
+                "{} - not handling message - already handled: {:?}",
+                self,
+                msg_with_bytes
+            );
             return Ok(Transition::Stay);
         }
 
-        let msg = msg.take_or_deserialize_message()?;
+        let msg = msg_with_bytes.take_or_deserialize_message()?;
+
         if self.should_handle_message(&msg) && self.verify_message(&msg)? {
+            self.set_message_handled(&msg_with_bytes);
```

Comment:
That won't happen because if `should_handle_message` returns true, the message is always handled and never bounced. But I'll make the change anyway, for sake of clarity.